crates/wasi-nn/src/backend/onnxruntime.rs

```diff
@@ -0,0 +1,182 @@
+    fn set_input(&mut self, index: u32, tensor: &Tensor) -> Result<(), BackendError> {
+        self.inputs[index as usize].replace(tensor.clone());
+        Ok(())
+    }
+
+    fn compute(&mut self) -> Result<(), BackendError> {
+        let shaped_inputs: Vec<_> = self
+            .inputs
+            .iter()
+            .enumerate()
+            .map(|(i, _o)| {
+                let input = self.inputs[i].as_ref().unwrap();
+                let dims = input
+                    .dimensions
+                    .as_slice()
+                    .iter()
+                    .map(|d| *d as usize)
+                    .collect::<Vec<_>>();
+                match input.tensor_type {
+                    TensorType::Fp32 => {
+                        let data = bytes_to_f32_vec(input.data.to_vec());
+                        inputs![Array::from_shape_vec(dims, data).unwrap()].unwrap()
+                    }
+                    TensorType::Fp16 => {
+                        unimplemented!("Fp16 not supported by ONNX");
+                    }
+                    TensorType::Bf16 => {
+                        unimplemented!("Bf16 not supported by ONNX");
+                    }
+                    TensorType::U8 => {
+                        unimplemented!("U8 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                    TensorType::I32 => {
+                        unimplemented!("I32 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                    TensorType::I64 => {
+                        unimplemented!("I64 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                    TensorType::Fp64 => {
+                        unimplemented!("Fp64 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                }
+            })
+            .flatten()
+            .collect();
+
+        let session = self.session.lock().unwrap();
+        let res = session.run(shaped_inputs.as_slice())?;
+
+        for i in 0..self.outputs.len() {
+            let raw: (Vec<i64>, &[f32]) = res[i].extract_raw_tensor()?;
+            let f32s = raw.1.to_vec();
+            self.outputs[i].replace(f32_vec_to_bytes(f32s));
+        }
+        Ok(())
+    }
+
+    fn get_output(&mut self, index: u32, destination: &mut [u8]) -> Result<u32, BackendError> {
+        let output = self.outputs[index as usize].as_ref().unwrap();
+        destination[..output.len()].copy_from_slice(output);
+        Ok(output.len() as u32)
+    }
+}
+
+impl From<ort::Error> for BackendError {
+    fn from(e: ort::Error) -> Self {
+        BackendError::BackendAccess(e.into())
+    }
+}
+
+pub fn f32_vec_to_bytes(data: Vec<f32>) -> Vec<u8> {
```

Comment:
Maybe we should explain somewhere why we are forced to copy the tensor twice &mdash; at input and output.